### ISSUE TITLE
Is it possible to add L1/L2 regularization using the trainer class ?

### ISSUE THREAD

**Issue by mayank64ce** [huggingface/transformers#33130]
### Feature request

I want to add L1/L2 regularization to the transformer training.

### Motivation

Adding L1/L2 reg can promote sparser models that can accelerate inference and reduce storage.

### Your contribution

Not sure.

**Comment by NielsRogge** [comment]
Hi,

If you want to add L1/L2 regularization to your loss function, you can overwrite the Trainer, as shown [here](https://discuss.huggingface.co/t/specify-loss-for-trainer-trainingarguments/10481/2?u=nielsr).

**Comment by mayank64ce** [comment]
Thanks for the reply. Is it possible to pass additional parameters (like regularization weights) to the `compute_loss` method ?

**Comment by NielsRogge** [comment]
No, I'm afraid adding those isn't possible, as we already support [114 flags](https://huggingface.co/docs/transformers/en/main_classes/trainer#transformers.TrainingArguments).

**Comment by qgallouedec** [comment]
Hey @mayank64ce, sorry for the very late reply, we're catching up on the backlog. L2 is already covered by the `weight_decay` training argument (decoupled AdamW weight decay, automatically skipped for biases and LayerNorm). For L1, `Trainer` has gained a `compute_loss_func` argument since this thread, so subclassing is no longer needed:

https://github.com/huggingface/transformers/blob/9c944e97629df226ea6fbfa9a615046082c96b09/src/transformers/trainer.py#L309-L311

It is called as `compute_loss_func(outputs, labels, num_items_in_batch=...)`, so close over the model and your regularization weight in the callable instead of passing them in as arguments. Closing since we are not planning dedicated regularization flags on top of that, but please re-open if that does not cover your case.
